File: 309-best-time-to-buy-and-sell-stock-with-cooldown/309-best-time-to-buy-and-sell-stock-with-cooldown.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        N = len(prices)
        @cache
        def dfs(i, buying):
            if i >= N: return 0
            cooldown = dfs(i+1, buying)
            if buying:
                buy = dfs(i+1, not buying) - prices[i]
                return max(buy, cooldown)
            else:
                sell = dfs(i+2, True) + prices[i]
                return max(sell, cooldown)
        
        return dfs(0, True)

File: 309-best-time-to-buy-and-sell-stock-with-cooldown/309-best-time-to-buy-and-sell-stock-with-cooldown.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int n = prices.size();
        vector<int> hasStock(n), noStock(n), justSold(n);
        hasStock[0] = -prices[0];
        justSold[0] = INT_MIN;
        for (int i = 1; i < n; i++) {
            noStock[i] = max(noStock[i - 1], justSold[i - 1]);
            hasStock[i] = max(noStock[i-1] - prices[i], hasStock[i - 1]);
            justSold[i] = hasStock[i-1] + prices[i];
        }
        return max(noStock.back(), justSold.back());
    }
};